libraries/chain/protocol/transaction.cpp

```diff
@@ -303,6 +303,15 @@ void verify_authority( const vector<operation>& ops, const flat_set<public_key_t
 
 
 flat_set<public_key_type> signed_transaction::get_signature_keys( const chain_id_type& chain_id )const
+{
+   // Strictly we should check whether the given chain ID is same as the one used to initialize the object.
+   // However, we don't pass in another chain ID so far, for better performance, we skip the check.
+   if( !signees.empty() || signatures.empty() )
+      return signees;
+   return _get_signature_keys( chain_id );
```

Comment:
I would swap the logic:
```
if( signees.empty() ) signees = _get_signature_keys;
return signees;
```

Reply:
I'd like to do that as well, but it breaks the `const` modifier, need to update quite some other code to adapt the change. Not returning a reference is due to the same reason.

Reply:
Ah. Hm, a slightly more complicated approach might be worth it...
https://stackoverflow.com/questions/3374743/c-const-getter-method-with-lazy-initialization

Reply:
Thanks for the link. I was thinking about using a `shared_ptr` or similar for the `signees` field. Looks like it's easier to use the `mutable` keyword?

Reply:
Yes (although slightly ugly).

Reply:
Update: after updated it to return a reference, a lot of test cases are broken due to reuse of `signed_transaction` objects. For example, construct a transaction, sign it, verify signatures, then clear the signatures, sign it with another key, verify signatures again -- this will fail because `signees` field is not refreshed. I think the main routine doesn't reuse `signed_transaction` objects so I'll fix the test cases.